// ngx-tweet/esm5/src/app/ngx-tweet.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgxTweetComponent } from './components/ngx-tweet.component';
import { NgxTweetService } from './services/ngx-tweet.service';
var NgxTweetModule = /** @class */ (function () {
    function NgxTweetModule() {
    }
    NgxTweetModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        NgxTweetComponent
                    ],
                    imports: [CommonModule],
                    exports: [NgxTweetComponent],
                    providers: [NgxTweetService]
                },] },
    ];
    return NgxTweetModule;
}());
export { NgxTweetModule };
function NgxTweetModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxTweetModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxTweetModule.ctorParameters;
}
