// ngx-tweet/bundles/ngx-tweet.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ngx-tweet', ['exports', '@angular/common', '@angular/core', 'rxjs'], factory) :
    (factory((global['ngx-tweet'] = {}),global.ng.common,global.ng.core,global.rxjs));
}(this, (function (exports,common,core,rxjs) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NgxTweetService = (function () {
        function NgxTweetService(_document) {
            this._document = _document;
            this.TWITTER_OBJECT = 'twttr';
            this.TWITTER_SCRIPT_ID = 'twitter-wjs';
            this.TWITTER_WIDGET_URL = 'https://platform.twitter.com/widgets.js';
        }
        /**
         * @return {?}
         */
        NgxTweetService.prototype.loadScript = /**
         * @return {?}
         */
            function () {
                var _this = this;
                return rxjs.Observable.create(function (observer) {
                    _this._startScriptLoad();
                    _this._document.defaultView[_this.TWITTER_OBJECT].ready(_this._onTwitterScriptLoadedFactory(observer));
                });
            };
        /**
         * @return {?}
         */
        NgxTweetService.prototype._startScriptLoad = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ twitterData = this._document.defaultView[this.TWITTER_OBJECT] || {};
                if (this._twitterScriptAlreadyExists()) {
                    this._document.defaultView[this.TWITTER_OBJECT] = twitterData;
                    return;
                }
                this._appendTwitterScriptToDOM();
                twitterData._e = [];
                twitterData.ready = function (callback) {
                    twitterData._e.push(callback);
                };
                this._document.defaultView[this.TWITTER_OBJECT] = twitterData;
            };
        /**
         * @return {?}
         */
        NgxTweetService.prototype._twitterScriptAlreadyExists = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ twitterScript = this._document.getElementById(this.TWITTER_SCRIPT_ID);
                return (twitterScript !== null || typeof twitterScript !== 'object');
            };
        /**
         * @return {?}
         */
        NgxTweetService.prototype._appendTwitterScriptToDOM = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ firstJSScript = this._document.getElementsByTagName('script')[0];
                var /** @type {?} */ js = this._document.createElement('script');
                js.id = this.TWITTER_SCRIPT_ID;
                js.src = this.TWITTER_WIDGET_URL;
                firstJSScript.parentNode.insertBefore(js, firstJSScript);
            };
        /**
         * @param {?} observer
         * @return {?}
         */
        NgxTweetService.prototype._onTwitterScriptLoadedFactory = /**
         * @param {?} observer
         * @return {?}
         */
            function (observer) {
                return function (twitterData) {
                    observer.next(twitterData);
                    observer.complete();
                };
            };
        NgxTweetService.decorators = [
            { type: core.Injectable },
        ];
        /** @nocollapse */
        NgxTweetService.ctorParameters = function () {
            return [
                { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] },] },
            ];
        };
        return NgxTweetService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NgxTweetComponent = (function () {
        function NgxTweetComponent(_elementRef, _ngxTweetService, _changeDetectorRef) {
            this._elementRef = _elementRef;
            this._ngxTweetService = _ngxTweetService;
            this._changeDetectorRef = _changeDetectorRef;
            this.isTwitterScriptLoading = true;
        }
        /**
         * @return {?}
         */
        NgxTweetComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this._loadTwitterScript();
            };
        /**
         * @return {?}
         */
        NgxTweetComponent.prototype._loadTwitterScript = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this._ngxTweetService
                    .loadScript()
                    .subscribe(function (twitterData) {
                    _this._updateTwitterScriptLoadingState();
                    twitterData.widgets.createTweet(_this.tweetId, _this._elementRef.nativeElement, {});
                });
            };
        /**
         * @return {?}
         */
        NgxTweetComponent.prototype._updateTwitterScriptLoadingState = /**
         * @return {?}
         */
            function () {
                this.isTwitterScriptLoading = false;
                this._changeDetectorRef.detectChanges();
            };
        NgxTweetComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-tweet',
                        template: '<ng-content *ngIf="isTwitterScriptLoading"></ng-content>',
                        styles: ['.twitter-tweet { transform: none !important; }'],
                        encapsulation: core.ViewEncapsulation.None,
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    },] },
        ];
        /** @nocollapse */
        NgxTweetComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef, },
                { type: NgxTweetService, },
                { type: core.ChangeDetectorRef, },
            ];
        };
        NgxTweetComponent.propDecorators = {
            "tweetId": [{ type: core.Input },],
        };
        return NgxTweetComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NgxTweetModule = (function () {
        function NgxTweetModule() {
        }
        NgxTweetModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            NgxTweetComponent
                        ],
                        imports: [common.CommonModule],
                        exports: [NgxTweetComponent],
                        providers: [NgxTweetService]
                    },] },
        ];
        return NgxTweetModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.NgxTweetModule = NgxTweetModule;
    exports.ɵa = NgxTweetComponent;
    exports.ɵb = NgxTweetService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
